### PR TITLE
Add DMA to the s32k1 serial

### PR DESCRIPTION
**Pull request by davids5**
## Summary

s32k1xx:serial Add EDMA

Fix code that would not compile

Add and option to use the GPIO with watermarks as RTS instead of the RTS on every character.

## Impact

CPU load is less at high Baud rates.

At 1Mbs, The per character interrupt rate was every 10 uS.   With this PR the RX DMA, with a buffer of 64, will interrupt the CPU every 319uS with an off load time of 71uS

## Testing

PX4 nxp_ucans32k146

### CODE REVIEW

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 1]
```diff
@@ -191,12 +241,46 @@ static int  s32k1xx_attach(struct uart_dev_s *dev);
...
+static int  s32k1xx_dma_receive(struct uart_dev_s *dev,
+                                unsigned int *status);
+#ifdef CONFIG_PM
+static void s32k1xx_dma_reenable(struct s32k1xx_uart_s *priv);
+#endif
+static void s32k1xx_dma_rxint(struct uart_dev_s *dev, bool enable);
+static bool s32k1xx_dma_rxavailable(struct uart_dev_s *dev);
+
+static void s32k1xx_dma_rxcallback(DMACH_HANDLE handle, void *arg, bool done,
+                                  int result);
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:281]:
```suggestion
static void s32k1xx_dma_rxcallback(DMACH_HANDLE handle, void *arg, bool done,
                                   int result);
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 2]
```diff
@@ -191,12 +241,46 @@ static int  s32k1xx_attach(struct uart_dev_s *dev);
...
+                                unsigned int nbuffered, bool upper);
+#endif
+static void s32k1xx_send(struct uart_dev_s *dev, int ch);
 static bool s32k1xx_txready(struct uart_dev_s *dev);
+#ifdef SERIAL_HAVE_TXDMA
+static void s32k1xx_dma_send(struct uart_dev_s *dev);
+static void s32k1xx_dma_txint(struct uart_dev_s *dev, bool enable);
+static void s32k1xx_dma_txavailable(struct uart_dev_s *dev);
+static void s32k1xx_dma_txcallback(DMACH_HANDLE handle, void *arg, bool done,
+                                  int result);
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:263]:
```suggestion
static void s32k1xx_dma_txcallback(DMACH_HANDLE handle, void *arg, bool done,
                                   int result);
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 3]
```diff
@@ -191,12 +241,46 @@ static int  s32k1xx_attach(struct uart_dev_s *dev);
...
 static void s32k1xx_rxint(struct uart_dev_s *dev, bool enable);
 static bool s32k1xx_rxavailable(struct uart_dev_s *dev);
-static void s32k1xx_send(struct uart_dev_s *dev, int ch);
+#endif
+#if !defined(SERIAL_HAVE_ONLY_TXDMA)
 static void s32k1xx_txint(struct uart_dev_s *dev, bool enable);
+#endif
+#ifdef CONFIG_SERIAL_IFLOWCONTROL
+static bool s32k1xx_rxflowcontrol(struct uart_dev_s *dev,
+                                unsigned int nbuffered, bool upper);
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:254]:
```suggestion
static bool s32k1xx_rxflowcontrol(struct uart_dev_s *dev,
                                  unsigned int nbuffered, bool upper);
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 4]
```diff
@@ -473,6 +709,132 @@ static inline void s32k1xx_restoreuartint(struct s32k1xx_uart_s *priv,
...
+
+#if defined(SERIAL_HAVE_TXDMA)
+  /* Acquire the Tx DMA channel.  This should always succeed. */
+
+  if (priv->dma_txreqsrc != 0)
+    {
+      if (priv->txdma == NULL)
+        {
+          priv->txdma = s32k1xx_dmach_alloc(priv->dma_txreqsrc |
+                                          DMAMUX_CHCFG_ENBL, 0);
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:749]:
```suggestion
          priv->txdma = s32k1xx_dmach_alloc(priv->dma_txreqsrc |
                                            DMAMUX_CHCFG_ENBL, 0);
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 5]
```diff
@@ -473,6 +709,132 @@ static inline void s32k1xx_restoreuartint(struct s32k1xx_uart_s *priv,
...
+
+#if defined(SERIAL_HAVE_RXDMA)
+  /* Acquire the Rx DMA channel.  This should always succeed. */
+
+  if (priv->dma_rxreqsrc != 0)
+    {
+      if (priv->rxdma == NULL)
+        {
+          priv->rxdma = s32k1xx_dmach_alloc(priv->dma_rxreqsrc |
+                                          DMAMUX_CHCFG_ENBL, 0);
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:774]:
```suggestion
          priv->rxdma = s32k1xx_dmach_alloc(priv->dma_rxreqsrc |
                                            DMAMUX_CHCFG_ENBL, 0);
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 6]
```diff
@@ -473,6 +709,132 @@ static inline void s32k1xx_restoreuartint(struct s32k1xx_uart_s *priv,
...
+        }
+      else
+        {
+          s32k1xx_dmach_stop(priv->rxdma);
+        }
+
+      /* Configure for circular DMA reception into the RX FIFO */
+
+      config.saddr  = priv->uartbase + S32K1XX_LPUART_DATA_OFFSET;
+      config.daddr  = (uint32_t) priv->rxfifo;
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:789]:
```suggestion
      config.daddr  = (uint32_t)priv->rxfifo;
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 7]
```diff
@@ -473,6 +709,132 @@ static inline void s32k1xx_restoreuartint(struct s32k1xx_uart_s *priv,
...
+      modifyreg32(priv->uartbase + S32K1XX_LPUART_BAUD_OFFSET,
+                  0, LPUART_BAUD_RDMAE);
+
+      /* Enable itnerrupt on Idel and erros */
+
+      modifyreg32(priv->uartbase + S32K1XX_LPUART_CTRL_OFFSET, 0,
+                  LPUART_CTRL_PEIE       |
+                  LPUART_CTRL_FEIE       |
+                  LPUART_CTRL_NEIE       |
+                  LPUART_CTRL_ILIE);
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:823]:
```suggestion
      modifyreg32(priv->uartbase + S32K1XX_LPUART_CTRL_OFFSET, 0,
                  LPUART_CTRL_PEIE |
                  LPUART_CTRL_FEIE |
                  LPUART_CTRL_NEIE |
                  LPUART_CTRL_ILIE);
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 8]
```diff
@@ -703,6 +1151,7 @@ static int s32k1xx_ioctl(struct file *filep, int cmd, unsigned long arg)
 #if defined(CONFIG_SERIAL_TIOCSERGSTRUCT) || defined(CONFIG_SERIAL_TERMIOS)
   struct inode *inode = filep->f_inode;
   struct uart_dev_s *dev   = inode->i_private;
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:1153]:
```suggestion
  struct uart_dev_s *dev = inode->i_private;
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 9]
```diff
@@ -877,13 +1326,15 @@ static int s32k1xx_ioctl(struct file *filep, int cmd, unsigned long arg)
              * implement TCSADRAIN / TCSAFLUSH
              */
 
+            flags  = spin_lock_irqsave(NULL);
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:1329]:
```suggestion
            flags = spin_lock_irqsave(NULL);
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 10]
```diff
@@ -1022,6 +1478,359 @@ static bool s32k1xx_rxavailable(struct uart_dev_s *dev)
...
+ *               false indicates the lower watermark has been crossed
+ *
+ * Returned Value:
+ *   true if RX flow control activated.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SERIAL_IFLOWCONTROL
+static bool s32k1xx_rxflowcontrol(struct uart_dev_s *dev,
+                             unsigned int nbuffered, bool upper)
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:1508]:
```suggestion
static bool s32k1xx_rxflowcontrol(struct uart_dev_s *dev,
                                  unsigned int nbuffered, bool upper)
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 11]
```diff
@@ -1022,6 +1478,359 @@ static bool s32k1xx_rxavailable(struct uart_dev_s *dev)
...
+ *
+ * Description:
+ *   This function clears dma buffer at complete of DMA transfer and wakes up
+ *   threads waiting for space in buffer.
+ *
+ ****************************************************************************/
+
+#ifdef SERIAL_HAVE_TXDMA
+static void s32k1xx_dma_txcallback(DMACH_HANDLE handle, void *arg, bool done,
+                                  int result)
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:1735]:
```suggestion
static void s32k1xx_dma_txcallback(DMACH_HANDLE handle, void *arg, bool done,
                                   int result)
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 12]
```diff
@@ -1022,6 +1478,359 @@ static bool s32k1xx_rxavailable(struct uart_dev_s *dev)
...
+  /* Make use of setup function to update buffer and its length for next
+   * transfer
+   */
+
+  config.iter   = dev->dmatx.length;
+  config.flags  = EDMA_CONFIG_LINKTYPE_LINKNONE;
+  config.ssize  = EDMA_8BIT;
+  config.dsize  = EDMA_8BIT;
+  config.nbytes = sizeof(dev->dmatx.buffer[0]);
+  config.saddr  = (uint32_t) dev->dmatx.buffer;
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:1807]:
```suggestion
  config.saddr  = (uint32_t)dev->dmatx.buffer;
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 13]
```diff
@@ -1022,6 +1478,359 @@ static bool s32k1xx_rxavailable(struct uart_dev_s *dev)
...
+  /* Setup first half */
+
+  s32k1xx_dmach_xfrsetup(priv->txdma, &config);
+
+  /* Is this a split transfer? */
+
+  if (dev->dmatx.nbuffer)
+    {
+      config.iter   = priv->dev.dmatx.nlength;
+      config.saddr  = (uint32_t) priv->dev.dmatx.nbuffer;
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:1824]:
```suggestion
      config.saddr  = (uint32_t)priv->dev.dmatx.nbuffer;
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 14]
```diff
@@ -1106,6 +1939,48 @@ static bool s32k1xx_txempty(struct uart_dev_s *dev)
...
+ *
+ * Description:
+ *   This function checks the current DMA state and calls the generic
+ *   serial stack when bytes appear to be available.
+ *
+ ****************************************************************************/
+
+#ifdef SERIAL_HAVE_RXDMA
+static void s32k1xx_dma_rxcallback(DMACH_HANDLE handle, void *arg, bool done,
+                                  int result)
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:1953]:
```suggestion
static void s32k1xx_dma_rxcallback(DMACH_HANDLE handle, void *arg, bool done,
                                   int result)
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.c [hunk 15]
```diff
@@ -1106,6 +1939,48 @@ static bool s32k1xx_txempty(struct uart_dev_s *dev)
...
+   */
+
+  sr = s32k1xx_serialin(priv, S32K1XX_LPUART_STAT_OFFSET);
+
+  if ((sr & (LPUART_STAT_OR | LPUART_STAT_NF | LPUART_STAT_FE)) != 0)
+    {
+      s32k1xx_serialout(priv, S32K1XX_LPUART_STAT_OFFSET,
+                      sr & (LPUART_STAT_OR |
+                            LPUART_STAT_NF |
+                            LPUART_STAT_FE));
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.c:1979]:
```suggestion
      s32k1xx_serialout(priv, S32K1XX_LPUART_STAT_OFFSET,
                        sr & (LPUART_STAT_OR |
                              LPUART_STAT_NF |
                              LPUART_STAT_FE));
```

File: arch/arm/src/s32k1xx/s32k1xx_serial.h [hunk 16]
```diff
@@ -34,6 +34,115 @@
...
+#endif
+
+/* Verify that DMA has been enabled and the DMA channel has been defined.
+ */
+
+#  if defined(SERIAL_HAVE_TXDMA) || defined(SERIAL_HAVE_RXDMA)
+#    ifndef CONFIG_S32K1XX_EDMA
+#      error S32K1XX LPUART receive or transmit DMA requires CONFIG_S32K1XX_EDMA
+#    endif
+#  endif
```
Comment (pkarashchenko) [inline arch/arm/src/s32k1xx/s32k1xx_serial.h:122]:
```suggestion
#if defined(SERIAL_HAVE_TXDMA) || defined(SERIAL_HAVE_RXDMA)
#  ifndef CONFIG_S32K1XX_EDMA
#    error S32K1XX LPUART receive or transmit DMA requires CONFIG_S32K1XX_EDMA
#  endif
#endif
```

**Comment by davids5** [comment]
@pkarashchenko - Thank you - I incorporated all the style changes. Do you have a tool you are using or just a great eye?

The build error was from not having s32k11x DMAMUX that should be resolved now.

**Comment by pkarashchenko** [comment]
> @pkarashchenko - Thank you - I incorporated all the style changes. Do you have a tool you are using or just a great eye?
> 
> The build error was from not having s32k11x DMAMUX that should be resolved now.

No tool. Just my eyes.

**Comment by davids5** [comment]
@xiaoxiang781216 Any input?

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**